File: mobile/lib/models/mart_bill.dart
import 'json_parsers.dart';

class MartBill {
  final int id;
  final String? invoiceDate;
  final int martId;
  final String? martName;
  final double totalAmount;
  final String? filePath;
  final String status;
  final String? remarks;
  final String? lockedAt;
  final String? lockedBy;

  MartBill({
    required this.id,
    this.invoiceDate,
    required this.martId,
    this.martName,
    required this.totalAmount,
    this.filePath,
    required this.status,
    this.remarks,
    this.lockedAt,
    this.lockedBy,
  });

  factory MartBill.fromJson(Map<String, dynamic> json) {
    return MartBill(
      id: JsonParsers.parseInt(json['id']),
      invoiceDate: json['invoice_date']?.toString(),
      martId: JsonParsers.parseInt(json['mart_id']),
      martName: json['mart_name']?.toString(),
      totalAmount: JsonParsers.parseNum(json['total_amount']),
      filePath: json['file_path']?.toString(),
      status: JsonParsers.parseString(json['status'], fallback: 'PROCESSING'),
      remarks: json['remarks']?.toString(),
      lockedAt: json['locked_at']?.toString(),
      lockedBy: json['locked_by']?.toString(),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'invoice_date': invoiceDate,
    'mart_id': martId,
    'mart_name': martName,
    'total_amount': totalAmount,
    'file_path': filePath,
    'status': status,
    'remarks': remarks,
    'locked_at': lockedAt,
    'locked_by': lockedBy,
  };
}

File: backend/app/db/models/mart_bill.py
from sqlalchemy import Column, Date, DateTime, ForeignKey, Integer, Numeric, String
from sqlalchemy.orm import relationship

from .base_class import Base
from .mixins import AuditMixin


class MartBill(Base, AuditMixin):
    __tablename__ = "invoice"

    id = Column(Integer, primary_key=True, index=True)
    mart_id = Column(Integer, ForeignKey("mart.id"), nullable=False)
    warehouse_id = Column(
        Integer, ForeignKey("warehouse.id"), nullable=False, index=True
    )
    invoice_date = Column(Date, nullable=False)
    file_path = Column(String, nullable=False)
    file_hash = Column(String, nullable=False, unique=True, index=True)
    total_amount = Column(Numeric(18, 6), nullable=True)
    # Lifecycle Status: 'PROCESSING', 'NEEDS_REVIEW', 'VERIFIED'
    status = Column(String, default="PROCESSING", nullable=False)
    locked_at = Column(DateTime, nullable=True)
    locked_by = Column(String, nullable=True)
    locked_by = Column(String, nullable=True)
    remarks = Column(String, nullable=True)
    format_type = Column(String, nullable=True)

    mart = relationship("Mart")
    warehouse = relationship("Warehouse")
    items = relationship(
        "MartBillItem", back_populates="bill", cascade="all, delete-orphan"
    )
